package cloudproxy;

enum Op {
    ALL = 0;
    CREATE = 1;
    DESTROY = 2;
    WRITE = 3;
    READ = 4;
}

message Action {
    required string subject = 1;
    required Op verb = 2;
    required string object = 3;
}

message Auth {
    required string subject = 1;
}

message ACL {
    repeated Action permissions = 1;
}

message SignedACL {
    required bytes serialized_acls = 1;
    required string signature = 2;
}

message Challenge {
    required string subject = 1;
    required string nonce = 2;
}

message Response {
    required bytes serialized_chall = 1;
    required string signature = 2;
}

message ClientMessage {
    optional Action action = 1;
    optional Auth auth = 2;
    optional Response response = 3;
}

message ServerMessage {
    optional Challenge challenge = 1;
}

// a message that gives a name to a given keyczar public key structure
message SpeaksFor {
    required string subject = 1;
    required string pub_key = 2;
    required string meta = 3;
}

message SignedSpeaksFor {
    required bytes serialized_speaks_for = 1;
    required string signature = 2;
}
